Try to import docker images before falling back to building

arlyon opened the pull request:
<!-- Include if applicable: -->
**Fixes**: #3886

**Description**

When perfoming a lookup of artifacts, this change will attempt to import artifacts from docker running locally as well as whatever remote registry is configured. This enables the worklow outlined in the issue above.

This PR is not done because there needs to be some discussion as it changes the behaviour of builds. Should the functionality be placed behind a flag? Since the contents of images aren't verified, this could for example cause problems if a malicious (or unaware) user uploads images tagged with an incorrect commit however for those who are aware of the (imo limited) drawbacks the productivity boost far-outweighs the risk. That said, I would argue to make it opt-in.

**User facing changes**

This would cause anyone running `skaffold build` to have the images be imported from docker, if possible, 'short circuiting' some builds. This is particularly useful when sharing images across teams (say someone needs to check out your branch to help diagnose an issue).

**Follow-up work**

Once the questions above are answered, then the PR needs to be adapted slightly and the docs updated.

arlyon commented:
@googlebot I signed it!

tstromberg commented on pkg/skaffold/build/cache/lookup.go:
```diff
@@ -59,7 +61,10 @@ func (c *cache) lookup(ctx context.Context, a *latest.Artifact, tag string) cach
-		return needsBuilding{hash: hash}
+		entry, err = c.tryImport(ctx, a, tag, hash)
+		if err != nil {
+			return needsBuilding{hash: hash}
```
It would be nice to log the error that caused it to still need building.

Otherwise, PR seems reasonable to me. Anything required to remove the 'Draft' label?

arlyon replied:
Nothing (aside from docs changes), as long as you're happy to make this the default behaviour, as opposed to opt in. I'll have a scan through the docs and see what needs updating.

arlyon commented:
@tstromberg Just fixed the failing test and added the debug log!

dgageot commented:
@arlyon I think it's best to make it opt-in or to make it possible to get the current behaviour back. I've always found it super important in a fully local mode to be able to never connect to a remote registry. For example, you are in a train, with no network, assuming all the layers where already downloaded, you are able to iterate quickly without reaching out to the network.

arlyon commented:
> @arlyon I think it's best to make it opt-in or to make it possible to get the current behaviour back. I've always found it super important in a fully local mode to be able to never connect to a remote registry. For example, you are in a train, with no network, assuming all the layers where already downloaded, you are able to iterate quickly without reaching out to the network.

You're right. For now I implemented it as a config option under local builds:

```yaml
build:
  artifacts:
  - image: arlyon/workout-server
    context: ./server/
  - image: arlyon/workout-learn
    context: ./learn/
  local:
    tryImportMissing: true
    push: true
```

tejal29 commented on pkg/skaffold/schema/latest/config.go:
```diff
@@ -191,6 +191,10 @@ type LocalBuild struct {
 
+	// TryImportMissing should attempt to import this artifact from
+	// Docker (either locally or a remote registry) if not in cache.
+	TryImportMissing bool `yaml:"tryImportMissing,omitempty"`
```
Can you please make this a pointer like other?
And then you would need to set to it false in `defaultToLocalBuild` in `pkg/skaffold/schemadefaults/`

tejal29 commented:
@arlyon are you still working on this?

nkubala commented:
@arlyon this PR is getting a bit stale but I'd love to help you get it in. let me know if you need any help with it! i'm still seeing a test failing on a data race in the cache somewhere, can you check that out?

arlyon commented:
Hey, haven't been using skaffold at my day job recently but I have some time today so I'll rebase and see if I can get the requested changes in today.

arlyon commented:
I have updated with the requested changes and fixed the potential data race issues by introducing a RWMutex. `make test` fails complaining that schema `v2beta5` is already released but I'm not sure what to do about that.

gsquared94 commented:
You can rebase on master now. A new schema version has already been created. That should help with the tests

arlyon commented:
Hate to ping because I know you're all busy but would appreciate a merge @gsquared94 @tejal29  :)

dgageot commented:
I can TAL tomorrow. Added to m'y TODO

Le ven. 24 juil. 2020 à 17:00, Alexander Lyon <notifications@github.com> a
écrit :

> Hate to ping because I know you're all busy but would appreciate a merge
> @gsquared94 <https://github.com/gsquared94> @tejal29
> <https://github.com/tejal29> :)
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleContainerTools/skaffold/pull/3891#issuecomment-663581756>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABFPF7Z4JYJ3KOXVKYIMCLR5GOYHANCNFSM4LXDX4IA>
> .
>

dgageot commented on pkg/skaffold/build/local/types.go:
```diff
@@ -75,12 +76,18 @@ func NewBuilder(runCtx *runcontext.RunContext) (*Builder, error) {
 
+	tryImportMissing := false
+	if runCtx.Cfg.Build.LocalBuild.TryImportMissing != nil {
+		tryImportMissing = *runCtx.Cfg.Build.LocalBuild.TryImportMissing
```
`TryImportMissing` should probably just be of type bool since it defaults to `false`

briandealwis replied:
@tejal29 asked him to turn it into a pointer.  A pointer at least allows us to later check if the value was explicitly specified.

dgageot replied:
What for?

briandealwis replied:
If we change the default?

dgageot replied:
I see. Not sure we should take such future changes into account. pointers on booleans are the worst

dgageot commented on pkg/skaffold/build/cache/lookup.go:
```diff
@@ -108,3 +116,39 @@ func (c *cache) lookupRemote(ctx context.Context, hash, tag string, entry ImageD
+		logrus.Debugf("Added digest for %s to cache entry", tag)
+		entry.Digest = digest
+	}
+	c.cacheMutex.Lock()
+	defer c.cacheMutex.Unlock()
+	c.artifactCache[hash] = entry
+	return entry, nil
```
```suggestion
	c.cacheMutex.Lock()
	c.artifactCache[hash] = entry
	c.cacheMutex.Unlock()
	return entry, nil
```

dgageot commented on pkg/skaffold/build/cache/cache.go:
```diff
@@ -80,10 +83,12 @@ func NewCache(runCtx *runcontext.RunContext, imagesAreLocal bool, dependencies D
 
 	return &cache{
 		artifactCache:      artifactCache,
+		cacheMutex:         sync.RWMutex{},
```
This line is not needed. The default value for `sync.RWMutex` works.

dgageot commented on pkg/skaffold/build/cache/details.go:
```diff
@@ -111,6 +111,8 @@ func (d needsPushing) Push(ctx context.Context, out io.Writer, c *cache) error {
 	}
 
 	// Update cache
+	c.cacheMutex.Lock()
+	defer c.cacheMutex.Unlock()
 	e := c.artifactCache[d.hash]
 	e.Digest = digest
 	c.artifactCache[d.hash] = e
```
```suggestion
	c.cacheMutex.Lock()
	e := c.artifactCache[d.hash]
	e.Digest = digest
	c.artifactCache[d.hash] = e
	c.cacheMutex.Unlock()
```

nkubala commented:
@arlyon this PR is getting a bit stale. do you still have the bandwidth to try and push it through?

arlyon commented:
Hi @nkubala sorry for the delay I though I'd pushed the requested changes but clearly not. I have rebased on master now!

MarlonGamez reviewed (approved):
@arlyon This looks good to me! Thanks for making all of the requested changes :)
